docs/instrumentation.md

```diff
@@ -0,0 +1,283 @@
+The rest of this patch is the code instrumentation functions themself.
+
+[instrument.patch](data/instrument.patch)
+
+**Build kernel**
+
+```shell
+cd git
+cd linux
+export KERNEL=kernel7
+# output build directories
+export BUILD_DIR=build
+export INSTALL_PATH=install_tmp
+# temp directory for new content of the /boot and /root partion 
+export BOOT_PART_TMP=$BUILD_DIR/$INSTALL_PATH/boot
+
+# make modules_install will create build/install_tmp/root/lib directory during
+# install. we dont need to specify $BUILD_DIR prefix
+export ROOT_PART_TMP=$INSTALL_PATH/root
+
+# prepare output directory
+mkdir -p $BOOT_PART_TMP
+mkdir -p $BOOT_PART_TMP/overlays
+
+make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- -j4 O=$BUILD_DIR bcm2709_defconfig
+make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- -j4 O=$BUILD_DIR scripts prepare modules_prepare
+make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- -j4 O=$BUILD_DIR zImage modules dtbs
+make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- -j4 O=$BUILD_DIR modules_install INSTALL_MOD_PATH=$ROOT_PART_TMP 
+
+# export K_VERSION variable
+export K_VERSION=$(make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- O=$BUILD_DIR -s kernelrelease)
+
+cp $BUILD_DIR/arch/arm/boot/zImage $BOOT_PART_TMP/$KERNEL.img
+
+cp $BUILD_DIR/arch/arm/boot/dts/*.dtb $BOOT_PART_TMP
+cp $BUILD_DIR/arch/arm/boot/dts/overlays/*.dtb $BOOT_PART_TMP/overlays
+```
+
+**Deploy**
+
+```shell
+export PI_HOME=/home/pi
+
+cd ~/git/linux
+tar -zcvf install_pkg.tar.gz $BUILD_DIR/$INSTALL_PATH
+scp install_pkg.tar.gz $RASPI_IP:/home/pi/
+
+ssh $RASPI_IP tar -xvzf $PI_HOME/install_pkg.tar.gz
+ssh $RASPI_IP rm $PI_HOME/install_pkg.tar.gz
+ssh $RASPI_IP sudo cp -r $PI_HOME/build/install_tmp/root/lib/* /lib
+ssh $RASPI_IP sudo cp -r $PI_HOME/build/install_tmp/boot/* /boot
+ssh $RASPI_IP rm -rf $PI_HOME/build
+ssh $RASPI_IP sudo reboot
+```
+
+Test if dynamic debug option is enabled
+
+```shell
+ssh $RASPI_IP sudo modprobe configs
+ssh $RASPI_IP "zcat /proc/config.gz | grep DYNAMIC_DEBUG"
+# you sould get the following output: 
+# CONFIG_DYNAMIC_DEBUG=y
+# CONFIG_DYNAMIC_DEBUG_CORE=y
+```
+
+**Build instrumented module**
+
+```shell
+cd ~/git/linux
+
+export BUILD_DIR=build
+export K_VERSION=$(make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- O=$BUILD_DIR -s kernelrelease)
+make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- -j4 O=$BUILD_DIR -C . M=drivers/char/tpm EXTRA_CFLAGS="-g -DDEBUG=1 -finstrument-functions"
+
+export MOD_SRC=build/drivers/char/tpm
+export MOD_TMP_DST=/home/pi
+export MOD_DST=/lib/modules/$K_VERSION/kernel/drivers/char/tpm
+
+scp $MOD_SRC/tpm.ko $RASPI_IP:$MOD_TMP_DST/
+scp $MOD_SRC/tpm_tis_core.ko $RASPI_IP:$MOD_TMP_DST
+scp $MOD_SRC/tpm_tis_spi.ko $RASPI_IP:$MOD_TMP_DST
+
+ssh $RASPI_IP sudo mv tpm.ko $MOD_DST;
+ssh $RASPI_IP sudo mv tpm_tis_core.ko $MOD_DST;
+ssh $RASPI_IP sudo mv tpm_tis_spi.ko $MOD_DST;
+```
+
+At this point, we need to enable device tree overlay for tpm.
+
+```shell
+sudo nano /boot/config.txt
```

Comment:
@acwiek It may be confusing for some which commands are executed on RPi and which on host.